oximeter/db/src/client.rs

```diff
@@ -267,14 +271,320 @@ impl Client {
+            return Err(Error::MissingSchemaVersion(latest));
+        }
+        if !available.contains(&desired_version) {
+            return Err(Error::MissingSchemaVersion(desired_version));
+        }
+
+        // Check we have no gaps in version numbers, starting with the latest
+        // version and walking through all available ones strictly greater. This
+        // is to check that the _next_ version is also 1 greater than the
+        // latest.
+        let range = (Bound::Excluded(latest), Bound::Included(desired_version));
+        if available
+            .range(latest..)
+            .zip(available.range(range))
+            .any(|(current, next)| next - current != 1)
+        {
+            return Err(Error::NonSequentialSchemaVersions);
+        }
+
+        // Walk through all changes between current version (exclusive) and
+        // the desired version (inclusive).
+        let versions_to_apply = available.range(range);
+        let mut current = latest;
+        for version in versions_to_apply {
+            if let Err(e) = self
+                .apply_one_schema_upgrade(replicated, *version, schema_dir)
+                .await
+            {
+                error!(
+                    self.log,
+                    "failed to apply schema upgrade";
+                    "current_version" => current,
+                    "next_version" => *version,
+                    "replicated" => replicated,
+                    "schema_dir" => schema_dir.display(),
+                    "error" => ?e,
+                );
+                return Err(e);
+            }
+            current = *version;
+            self.insert_version(current).await?;
+        }
+        Ok(())
+    }
+
+    fn verify_schema_upgrades(
+        files: &BTreeMap<String, (PathBuf, String)>,
+    ) -> Result<(), Error> {
+        let re =
+            regex::Regex::new("(INSERT INTO)|(ALTER TABLE .* DELETE)").unwrap();
+        for (path, sql) in files.values() {
+            if re.is_match(&sql) {
+                return Err(Error::SchemaUpdateModifiesData {
+                    path: path.clone(),
+                    statement: sql.clone(),
+                });
+            }
+            if sql.matches(';').count() > 1 {
+                return Err(Error::MultipleSqlStatementsInSchemaUpdate {
+                    path: path.clone(),
+                });
+            }
+        }
+        Ok(())
+    }
+
+    async fn apply_one_schema_upgrade(
+        &self,
+        replicated: bool,
+        next_version: u64,
+        schema_dir: impl AsRef<Path>,
+    ) -> Result<(), Error> {
+        let schema_dir = schema_dir.as_ref();
+        let upgrade_file_contents = Self::read_schema_upgrade_sql_files(
+            &self.log,
+            replicated,
+            next_version,
+            schema_dir,
+        )
+        .await?;
+
+        // We need to be pretty careful at this point with any data-modifying
```

Comment:
What would be wrong with inserting a placeholder row? I basically did this to insert the bootstrap "version" of the database

Reply:
I'm not sure I understand your question. Are you asking why we're validating anything about the queries? Or specifically about inserting a row in the version table?

Reply:
Kinda, yeah, like, what would be wrong with a data-modifying statement to "insert one row"?

Reply:
I think the short answer is that it's not clear to me one could do that and still have a valid set of data in the tables.

The `oximeter` collector normalizes the samples it receives, and always inserts records into at least 2 tables for any sample (at least one field, and the measurement). I think it's important for `oximeter` to be responsible for that, to ensure that its invariants are upheld. The version table is managed by the upgrade process, so I'm not sure we need to allow that now. The schema table is also managed automatically by `oximeter`, as it derives the schema from each sample.

I've been thinking about how to manage things like adding new columns that do require generating data. In those cases, I think we can relax this to adding something like a default expression to the `ADD COLUMN` statement. It's also possible that we want to _move_ data around, which would involve something like `INSERT INTO`. We can relax that when we need to.